# Springboard JSON Mini-Project

This mini project has 1 main purpose and 3 main objectives. 

## Purpose:

Enhance my knowledge using Python's Pandas and Json packages by exploring data from the World Bank

## Objectives:

With the JSON file provided I will:

1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using the column 'mjtheme_namecode')
3. Create a dataframe with the missing names filled in.



## Objective 1: Finding the top 10 Countries with the most projects

My approach for this objective would be to:
+ import the correct packages
+ import the Json data into a Pandas Dataframe
+ group the data by countries
+ use the agg function to count the number of project names
+ sort values in decending order by count and display the top 10


In [74]:
import numpy as np 
import pandas as pd
import json
from IPython.display import display 

#import the Json data into a Pandas Dataframe    
df = pd.read_json("world_bank_projects.json")

#group data by countries 
df_projects = df.groupby(["countryshortname"],as_index = False)["project_name"].agg("count")\
.rename(columns = {"countryshortname":"Country", "project_name": "Total Number of Projects"})

#sort values in decending order by count
df_top10_projects = df_projects.sort_values(by = "Total Number of Projects", ascending = False)
top10_projects = df_top10_projects.iloc[0:10,[0]].to_string(index = False)

display(df_top10_projects.head(10))
print("\nThe Countries with the top ten projects are:\n{}".format(top10_projects))

Country  Total Number of Projects
23                China                        19
45            Indonesia                        19
112             Vietnam                        17
44                India                        16
115  Yemen, Republic of                        13
73                Nepal                        12
8            Bangladesh                        12
69              Morocco                        12
70           Mozambique                        11
1                Africa                        11


The Countries with the top ten projects are:
Country
             China
         Indonesia
           Vietnam
             India
Yemen, Republic of
             Nepal
        Bangladesh
           Morocco
        Mozambique
            Africa


## Objective 2: Find the top 10 major project themes (using the column 'mjtheme_namecode')

My approach for this objective would be to:

+ Extract the mjtheme_namecode and mjthemecode column for all unique values
+ group the data by project name and use the agg function to count the number of project themes
+ sort values in decending order by count and display the top 10

In [62]:
# Extract the mjtheme_namecode and mjthemecode column for all unique values
import json
from pandas.io.json import json_normalize

with open("world_bank_projects.json") as file:
    json_data = json.load(file)
theme_codes = json_normalize(json_data,"mjtheme_namecode")

# group the data by project name and use the agg function to count the number of project themes
df_project_themes = theme_codes.groupby("name",as_index = False)["code"].agg("count")\
.rename(columns = {"name":"Project Name","code":"Total Number of Projects"})

# sort values in decending order by count and display the top 10
df_top10_project_themes = df_project_themes.sort_values(by = "Total Number of Projects", ascending = False)
df_top10_project_themes.head(10)


Project Name  Total Number of Projects
2   Environment and natural resources management                       223
7                              Rural development                       202
4                              Human development                       197
5                       Public sector governance                       184
9          Social protection and risk management                       158
3       Financial and private sector development                       130
0                                                                      122
8                    Social dev/gender/inclusion                       119
10                         Trade and integration                        72
11                             Urban development                        47

The Total Number of Project Themes shown in the cell above this is counting blank spaces as well. Looking where the index is 0, you can see that **122** of the Total Number of Projects are blank. 

## Objective 3: Create a dataframe with the missing names from objective 2 filled in

+ sort the values of the dataframe by the code and name
+ fill the blank values with Nan
+ bfill the NaN values

In [80]:
# sort the values of the dataframe by the code and name
theme_codes_sorted = theme_codes.sort_values(["code","name"])
display(theme_codes_sorted.head(10))

# fill the blank values with Nan
for name in theme_codes_sorted.loc[(theme_codes_sorted["name"] == "")]:
    theme_codes_sorted.loc[(theme_codes_sorted["name"] == ""),"name"] = np.nan
display(theme_codes_sorted.head(10))
    
#bfill the NaN values 
theme_codes_filled = theme_codes_sorted.fillna(method = "bfill")
display(theme_codes_filled.head(10))
display(theme_codes_filled.tail(10))

code                 name
212     1                     
363     1                     
1024    1                     
1114    1                     
1437    1                     
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management

code                 name
212     1                  NaN
363     1                  NaN
1024    1                  NaN
1114    1                  NaN
1437    1                  NaN
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management

code                 name
212     1  Economic management
363     1  Economic management
1024    1  Economic management
1114    1  Economic management
1437    1  Economic management
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management

code               name
1303    9  Urban development
1318    9  Urban development
1342    9  Urban development
1347    9  Urban development
1425    9  Urban development
1426    9  Urban development
1428    9  Urban development
1470    9  Urban development
1473    9  Urban development
1495    9  Urban development

## Objective 2 Continued: Find the top 10 major project themes (using the column 'mjtheme_namecode')
My approach for this objective would be to:

+ group the data by project name and use the agg function to count the number of project themes
+ sort values in decending order by count and display the top 10

The cell below shows the "Total Number of Projects" grouped by "Project Name" after all of the blank values
are filled in. 



In [81]:
# group the data by project name and use the agg function to count the number of project themes
df_project_themes = theme_codes_filled.groupby("name",as_index = False)["code"].agg("count")\
.rename(columns = {"name":"Project Name","code":"Total Number of Projects"})

# sort values in decending order by count and display the top 10
df_top10_project_themes = df_project_themes.sort_values(by = "Total Number of Projects", ascending = False)
df_top10_project_themes.head(10)

Project Name  Total Number of Projects
1   Environment and natural resources management                       250
6                              Rural development                       216
3                              Human development                       210
4                       Public sector governance                       199
8          Social protection and risk management                       168
2       Financial and private sector development                       146
7                    Social dev/gender/inclusion                       130
9                          Trade and integration                        77
10                             Urban development                        50
0                            Economic management                        38